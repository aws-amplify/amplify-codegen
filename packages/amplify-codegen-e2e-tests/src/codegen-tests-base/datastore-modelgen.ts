import {
    initProjectWithProfile,
    addApiWithSchemaAndConflictDetection,
    generateModels,
    AmplifyFrontendConfig
} from "amplify-codegen-e2e-core";
import { existsSync } from "fs";
import path from 'path';
import { isNotEmptyDir, generateSourceCode } from '../utils';

export async function testCodegenModels(config: AmplifyFrontendConfig, projectRoot: string, schema: string) {
    // init project and add API category
    await initProjectWithProfile(projectRoot, { ...config });

    //enable datastore
    await addApiWithSchemaAndConflictDetection(projectRoot, schema);

    //generate pre existing user file
    const userSourceCodePath = generateSourceCode(projectRoot, config.srcDir);

    //generate models
    await expect(generateModels(projectRoot)).resolves.not.toThrow();

    // pre-existing file should still exist
    expect(existsSync(userSourceCodePath)).toBe(true);
    // datastore models are generated at correct location
    expect(isNotEmptyDir(path.join(projectRoot, config.modelgenDir))).toBe(true);
}