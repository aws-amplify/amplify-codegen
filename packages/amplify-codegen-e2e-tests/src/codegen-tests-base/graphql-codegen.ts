import {
    initProjectWithProfile,
    addApiWithSchema,
    addCodegen,
    AmplifyFrontendConfig,
    generateStatementsAndTypes
} from "amplify-codegen-e2e-core";
import { existsSync } from "fs";
import path from 'path';
import { isNotEmptyDir, generateSourceCode } from '../utils';

export async function testGraphQLCodegen(config: AmplifyFrontendConfig, projectRoot: string, schema: string) {
    // init project and add API category
    await initProjectWithProfile(projectRoot, { ...config });
    await addApiWithSchema(projectRoot, schema);

    // generate pre-existing user file
    const userSourceCodePath = generateSourceCode(projectRoot, config.srcDir);

    // Add codegen
    await expect(addCodegen(projectRoot, { ...config })).resolves.not.toThrow();

    // Execute GraphQL codegen
    await expect(generateStatementsAndTypes(projectRoot)).resolves.not.toThrow();

    // check if the pre-existing user file still exists
    expect(existsSync(userSourceCodePath)).toBe(true);
    // check if the statements are generated
    expect(isNotEmptyDir(path.join(projectRoot, config.graphqlCodegenDir))).toBe(true);
}
