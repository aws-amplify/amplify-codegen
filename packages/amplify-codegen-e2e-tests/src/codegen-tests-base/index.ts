export * from './test-setup';
export * from './add-codegen';
export * from './graphql-codegen';
export * from './configure-codegen';
export * from './remove-codegen';
export * from './datastore-modelgen';
export * from './push-codegen';
