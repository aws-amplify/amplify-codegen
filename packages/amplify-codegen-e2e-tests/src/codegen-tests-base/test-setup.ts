import { 
    deleteProjectDir,
    deleteProject,
    AmplifyFrontendConfig,
    getAdminApp
} from "amplify-codegen-e2e-core";
import { existsSync, readFileSync } from "fs";
import path from 'path';
import { isNotEmptyDir, generateSourceCode } from '../utils';
import { JSONUtilities } from 'amplify-cli-core';
import { SandboxApp } from '../types/SandboxApp';

export function getGraphQLConfigFilePath(projectRoot: string): string {
    return path.join(projectRoot, '.graphqlconfig.yml');
}

export async function deleteAmplifyProject(projectRoot: string) {
    const metaFilePath = path.join(projectRoot, 'amplify', '#current-cloud-backend', 'amplify-meta.json');
      if (existsSync(metaFilePath)) {
        await deleteProject(projectRoot);
      }
      deleteProjectDir(projectRoot);
}

export function testSetupBeforeAddCodegen(projectRoot: string, config: AmplifyFrontendConfig): string {
    // generate pre-existing user file
    const userSourceCodePath = generateSourceCode(projectRoot, config.srcDir);
    
    // verify that the pre-existing file exists and codegen output directory is empty
    expect(existsSync(userSourceCodePath)).toBe(true);
    expect(isNotEmptyDir(path.join(projectRoot, config.graphqlCodegenDir))).toBe(false);
    
    return userSourceCodePath;
}

export async function testValidGraphQLConfig(projectRoot: string) {
    // graphql codegen configuration should exist
    expect(existsSync(getGraphQLConfigFilePath(projectRoot))).toBe(true);
    
    // check if the graphql codegen configuration is valid
    expect(readFileSync(getGraphQLConfigFilePath(projectRoot)).toString()).toMatchSnapshot();
}

export async function testSetupAdminApp(schemaBody: any = {}): Promise<string> {
    const sandBoxAppString = await getAdminApp(schemaBody);
    expect(sandBoxAppString).toBeDefined();
    
    const sandboxApp = JSONUtilities.parse<SandboxApp>(sandBoxAppString);
    expect(sandboxApp.schema).toEqual(schemaBody.schema);
    
    const sandboxId = sandboxApp.backendManagerAppId;
    expect(sandboxId).toBeDefined();
    return sandboxId;
}

