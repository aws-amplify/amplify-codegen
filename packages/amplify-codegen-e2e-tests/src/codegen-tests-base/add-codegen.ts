import {
    initProjectWithProfile,
    addApiWithSchema,
    addCodegen,
    AmplifyFrontendConfig
} from "amplify-codegen-e2e-core";
import { existsSync } from "fs";
import path from 'path';
import { isNotEmptyDir } from '../utils';
import { testSetupBeforeAddCodegen, testValidGraphQLConfig } from "./test-setup";

export async function testAddCodegen(config: AmplifyFrontendConfig, projectRoot: string, schema: string) {
    // init project and add API category
    await initProjectWithProfile(projectRoot, { ...config });
    await addApiWithSchema(projectRoot, schema);

    const userSourceCodePath = testSetupBeforeAddCodegen(projectRoot, config);

    // add codegen succeeds
    await expect(addCodegen(projectRoot, { ...config })).resolves.not.toThrow();

    // pre-existing file should still exist
    expect(existsSync(userSourceCodePath)).toBe(true);
    // GraphQL statements are generated
    expect(isNotEmptyDir(path.join(projectRoot, config.graphqlCodegenDir))).toBe(true);
    testValidGraphQLConfig(projectRoot, config);
}
