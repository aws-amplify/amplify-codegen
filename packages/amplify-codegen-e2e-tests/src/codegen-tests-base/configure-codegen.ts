import {
    initProjectWithProfile,
    addApiWithSchema,
    addCodegen,
    configureCodegen,
    AmplifyFrontendConfig
} from "amplify-codegen-e2e-core";
import { existsSync, readFileSync } from "fs";
import path from 'path';
import { isNotEmptyDir } from '../utils';
import { testSetupBeforeAddCodegen, testValidGraphQLConfig, getGraphQLConfigFilePath } from "./test-setup";

export async function testConfigureCodegen(config: AmplifyFrontendConfig, projectRoot: string, schema: string) {
    // init project and add API category
    await initProjectWithProfile(projectRoot, { ...config });
    await addApiWithSchema(projectRoot, schema);

    const userSourceCodePath = testSetupBeforeAddCodegen(projectRoot, config);

    // add codegen succeeds
    await expect(addCodegen(projectRoot, { ...config })).resolves.not.toThrow();

    const addedCodegenConfiguration = readFileSync(getGraphQLConfigFilePath(projectRoot)).toString();

    // update codegen configuration
    const settings = { isCodegenConfigured: true, maxStatementDepth: 4, ...config };
    await expect(configureCodegen(projectRoot, settings)).resolves.not.toThrow();

    // pre-existing file should still exist
    expect(existsSync(userSourceCodePath)).toBe(true);
    // previously generated files should still exist
    expect(isNotEmptyDir(path.join(projectRoot, config.graphqlCodegenDir))).toBe(true);

    // graphql configuration should be updated to with MaxStatementDepth=4
    testValidGraphQLConfig(projectRoot, config, 4, true);
    const updatedCodegenConfiguration = readFileSync(getGraphQLConfigFilePath(projectRoot)).toString();
    // the codegen configuration is updated
    expect(addedCodegenConfiguration).not.toMatch(updatedCodegenConfiguration);
}