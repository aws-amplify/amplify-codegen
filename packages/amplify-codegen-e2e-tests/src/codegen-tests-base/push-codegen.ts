import {
    initProjectWithProfile,
    addApiWithSchema,
    amplifyPushWithCodegenAdd,
    AmplifyFrontendConfig,
    apiUpdateToggleDataStore,
    amplifyPushWithCodegenUpdate
} from "amplify-codegen-e2e-core";
import { existsSync } from "fs";
import path from 'path';
import { isNotEmptyDir } from '../utils';
import { testSetupBeforeAddCodegen } from "./test-setup";

export async function testPushCodegen(config: AmplifyFrontendConfig, projectRoot: string, schema: string) {
    // init project and add API category
    await initProjectWithProfile(projectRoot, { ...config });
    await addApiWithSchema(projectRoot, schema);

    const userSourceCodePath = testSetupBeforeAddCodegen(projectRoot, config);

    // add codegen succeeds
    await amplifyPushWithCodegenAdd(projectRoot, { ...config });

    // pre-existing file should still exist
    expect(existsSync(userSourceCodePath)).toBe(true);

    // GraphQL statements are generated
    expect(isNotEmptyDir(path.join(projectRoot, config.graphqlCodegenDir))).toBe(true);

    //enable datastore
    await apiUpdateToggleDataStore(projectRoot);
    //push with codegen update
    await amplifyPushWithCodegenUpdate(projectRoot);
    expect(existsSync(userSourceCodePath)).toBe(true);
    expect(isNotEmptyDir(path.join(projectRoot, config.modelgenDir))).toBe(true);
}
