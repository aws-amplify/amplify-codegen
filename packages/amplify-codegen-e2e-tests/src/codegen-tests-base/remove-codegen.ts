import {
    initProjectWithProfile,
    addApiWithSchema,
    addCodegen,
    removeCodegen,
    AmplifyFrontendConfig
} from "amplify-codegen-e2e-core";
import { existsSync, readFileSync } from "fs";
import path from 'path';
import { isNotEmptyDir } from '../utils';
import { testSetupBeforeAddCodegen, getGraphQLConfigFilePath } from "./test-setup";
import { load } from 'js-yaml';

export async function testRemoveCodegen(config: AmplifyFrontendConfig, projectRoot: string, schema: string) {
    // init project and add API category
    await initProjectWithProfile(projectRoot, { ...config });
    await addApiWithSchema(projectRoot, schema);

    const userSourceCodePath = testSetupBeforeAddCodegen(projectRoot, config);

    // add codegen succeeds
    await expect(addCodegen(projectRoot, { ...config })).resolves.not.toThrow();

    // remove codegen
    await expect(removeCodegen(projectRoot)).resolves.not.toThrow();

    // pre-existing file should still exist
    expect(existsSync(userSourceCodePath)).toBe(true);
    // previously generated files should still exist
    expect(isNotEmptyDir(path.join(projectRoot, config.graphqlCodegenDir))).toBe(true);
    // graphql configuration should be updated to remove previous configuration
    const generatedConfig = load(readFileSync(getGraphQLConfigFilePath(projectRoot)).toString());
    expect(Object.keys(generatedConfig.projects).length).toEqual(0);
}