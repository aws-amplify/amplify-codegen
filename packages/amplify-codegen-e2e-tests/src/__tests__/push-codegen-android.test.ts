import { 
    createNewProjectDir,
    DEFAULT_ANDROID_CONFIG
} from "amplify-codegen-e2e-core";
import { deleteAmplifyProject, testPushCodegen } from '../codegen-tests-base';

const schema = 'simple_model.graphql';

describe('Amplify push with codegen tests - Android', () => {
    let projectRoot: string;
    beforeEach(async () => {
        projectRoot = await createNewProjectDir('pushCodegenAndroid');
    });

    afterEach(async () => {
        await deleteAmplifyProject(projectRoot); 
    });

    it(`should prompt codegen add/update and not delete user files`, async () => {
        await testPushCodegen(DEFAULT_ANDROID_CONFIG, projectRoot, schema);
    });
});