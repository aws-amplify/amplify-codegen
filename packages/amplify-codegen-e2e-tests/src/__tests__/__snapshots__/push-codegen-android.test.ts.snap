// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Amplify push with codegen tests - Android should prompt codegen add/update and not delete user files 1`] = `
"projects:
  pushcodegenandroidf0:
    schemaPath: app/src/main/graphql/schema.json
    includes:
      - app/src/main/graphql/**/*.graphql
    excludes:
      - ./amplify/**
    extensions:
      amplify:
        codeGenTarget: ''
        generatedFileName: ''
        docsFilePath: app/src/main/graphql/com/amazonaws/amplify/generated/graphql
"
`;
